// linkinator/build/src/queue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Queue = void 0;
const p_queue_1 = require("p-queue");
class Queue {
    constructor(options) {
        this.activeTimers = 0;
        this.q = new p_queue_1.default({
            concurrency: options.concurrency,
        });
    }
    add(fn, options) {
        if (options === null || options === void 0 ? void 0 : options.delay) {
            setTimeout(() => {
                this.q.add(fn);
                this.activeTimers--;
            }, options.delay);
            this.activeTimers++;
        }
        else {
            this.q.add(fn);
        }
    }
    async onIdle() {
        await this.q.onIdle();
        await new Promise(resolve => {
            if (this.activeTimers === 0) {
                resolve();
                return;
            }
            const timer = setInterval(async () => {
                if (this.activeTimers === 0) {
                    await this.q.onIdle();
                    clearInterval(timer);
                    resolve();
                    return;
                }
            }, 500);
        });
    }
}
exports.Queue = Queue;
